Frameworks/UIKit/UIViewController.mm

```diff
@@ -1168,27 +1164,72 @@ - (void)presentViewController:(UIViewController*)controller animated:(BOOL)anima
         return;
     }
 
-    if (controller != nil) {
-        [[UIApplication sharedApplication] beginIgnoringInteractionEvents];
+    UIViewController* visibleParent = self;
+    bool shouldShow = false;
+    do {
+        // Allow presentation only if an ancestor is not NotVisible.
+        shouldShow = visibleParent->priv->_visibility != controllerNotVisible;
+    } while (!shouldShow && (visibleParent = [visibleParent parentViewController]));
+
+    if (!shouldShow) {
+        TraceWarning(TAG, L"Controller is not visible!");
+        return;
+    }
+
+    if (visibleParent->priv->_visibility != controllerVisible) {
+        TraceWarning(TAG,
+                     L"Presenting view controller %08x (%hs) on view controller %08x (%hs) which is not yet fully visible - continuing anyway.",
+                     controller,
+                     object_getClassName(controller),
+                     visibleParent,
+                     object_getClassName(visibleParent));
     }
 
+    if (visibleParent != self) {
+        TraceWarning(TAG,
+                     L"View controller %08x (%hs) not visible but ancestor %08x (%hs) is - continuing anyway.",
+                     self,
+                     object_getClassName(self),
+                     visibleParent,
+                     object_getClassName(visibleParent));
+    }
+
+    [[UIApplication sharedApplication] beginIgnoringInteractionEvents];
+
     if ([controller _hidesParent]) {
         [self _notifyViewWillDisappear:animated];
     }
 
-    priv->_modalViewController = controller;
     priv->_presentedViewController = controller;
-    if (controller != nil) {
-        [controller view];
-        controller->priv->_parentViewController = self;
-        controller->priv->_presentingViewController = self;
-        controller->priv->_presentCompletionBlock.attach([completion copy]);
+    [controller view];
+    controller->priv->_parentViewController = self;
+    controller->priv->_presentingViewController = self;
 
-        if ([controller modalPresentationStyle] == UIModalPresentationPopover) {
-            controller->priv->_popoverPresentationController.attach([[UIPopoverPresentationController alloc] initWithPresentedViewController:controller presentingViewController:self]);
-        }
+    if ([controller modalPresentationStyle] == UIModalPresentationPopover) {
+        controller->priv->_popoverPresentationController.attach([[UIPopoverPresentationController alloc] initWithPresentedViewController:controller presentingViewController:self]);
+
+        dispatch_block_t popoverPresent = ^{
+            [[UIApplication sharedApplication] endIgnoringInteractionEvents];
 
-        [controller performSelectorOnMainThread:@selector(_addToTop:) withObject:[NSNumber numberWithInt:animated] waitUntilDone:NO];
+            if (completion) {
+                completion();
+            }
```

Comment:
right - the weakSelf pattern is to avoid cycles - probably ok to leave as is.  i hadn't realized that block capture copied block variables - thanks for clarifiying.  i'll keep my eye out for your last change, but that means i'll have to kick off another CI build - not a problem though.

---
In reply to: [95633496](https://github.com/Microsoft/WinObjC/pull/1601#discussion_r95633496) [](ancestors = 95633496)